src/mod/endpoints/mod_sofia/sofia.c

```diff
@@ -1516,6 +1516,14 @@ static void our_sofia_event_callback(nua_event_t event,
 
 	if (sofia_private && sofia_private != &mod_sofia_globals.destroy_private && sofia_private != &mod_sofia_globals.keep_private) {
 		if (!zstr(sofia_private->gateway_name)) {
+			if (event == nua_r_unregister && status == 200) {
```

Comment:
One issue if the registrar didn't respond to any final response. Check if we got like a timeout so we can also delete the nua handle. You can try to replicate it by creating a gateway then use sipp to act as a registrar. Stop responding to the UNREGISTER and check how this new implementation behave.

Reply:
Also check the scenario when user relogin immediately during unregister. Might cause a race condition where it deletes the nua handle or might cause memory leak since registering creates a new nua handle. You can try to emulate this with sipp.